swim/memberlist.go

```diff
@@ -381,47 +382,81 @@ func (m *memberlist) SetLocalLabels(labels map[string]string) error {
 // and subsequently be gossiped around. It is a valid operation to remove non-
 // existing keys. It returns true if all (and only all) labels have been removed.
 func (m *memberlist) RemoveLocalLabels(keys ...string) bool {
-	m.members.Lock()
-	if len(m.local.Labels) == 0 || len(keys) == 0 {
-		m.members.Unlock()
-		// nothing to delete
-		return false
-	}
+	// keep track if all labels are removed, it will be set to false if a label
+	// couldn't be removed.
+	removed := true
+
+	m.updateLocalMember(func(member *Member) bool {
+		// ensure that there is a new copy of the labels to work
+		// with.
+		labelsCopy := member.Labels.copy()
+
+		any := false // keep track if we at least removed one label
+		for _, key := range keys {
+			_, has := labelsCopy[key]
+			delete(labelsCopy, key)
+			removed = removed && has
+			any = any || has
+		}
 
-	any := false    // keep track if we at least removed one label
-	removed := true // keep track if all labels are removed
-	for _, key := range keys {
-		_, has := m.local.Labels[key]
-		delete(m.local.Labels, key)
-		removed = removed && has
-		any = any || has
-	}
+		if any {
+			// only if there are changes we put the copied labels on the member.
+			member.Labels = labelsCopy
+		}
 
-	m.members.Unlock()
+		return any
+	})
 
-	if any {
-		// only reincarnate if there is a label removed
-		m.postLocalUpdate()
-	}
 	return removed
 }
 
-// postLocalUpdate should be called after the local Member has been updated to
-// make sure that its new state has a higher incarnation number and the change
-// will be recorded as a change to gossip around.
-func (m *memberlist) postLocalUpdate() {
-	// bump our incarnation for this change to be accepted by all peers
+// updateLocalMember takes an update function to upate the member passed in. The
+// update function can make mutations to the member and should indicate if it
+// made changes, only if changes are made the incarnation number will be bumped
+// and the new state will be gossiped to the peers
+func (m *memberlist) updateLocalMember(update func(*Member) bool) {
+	var changes []Change
```

Comment:
Why is this moved up? I don't find this more readable. Having it in this form is fine.

I thought we had the `changes := []Change{}` form because it is marshaled into "[]" by the json pkg. While `var changes []Cange` will be Marshaled as "null"